src/System.Management.Automation/engine/NativeCommandParameterBinder.cs

```diff
@@ -324,12 +340,76 @@ private void PossiblyGlobArg(string arg, bool usedQuotes)
             }
 #endif // UNIX
 
+            // Check if the start of the path is a filesystem drive
+            if (stringConstantType != StringConstantType.SingleQuoted)
+            {
+                arg = ResolvePath(arg, Context);
+            }
+
             if (!argExpanded)
             {
                 _arguments.Append(arg);
             }
         }
 
+        /// <summary>
+        /// Check if string is prefixed by psdrive, if so, expand it if filesystem path.
+        /// </summary>
+        /// <param name="path">The potential PSPath to resolve.</param>
+        /// <param name="context">The current ExecutionContext.</param>
+        /// <returns>Resolved PSPath if applicable otherwise the original path</returns>
+        internal static string ResolvePath(string path, ExecutionContext context)
+        {
+            if (ExperimentalFeature.IsEnabled("PSNativePSPathResolution"))
+            {
+                string driveName;
+#if !UNIX
+                if (string.Equals(path, "~") || path.StartsWith("~" + Path.DirectorySeparatorChar) || path.StartsWith("~" + Path.AltDirectorySeparatorChar))
+                {
+                    try
+                    {
+                        ProviderInfo fileSystemProvider = context.EngineSessionState.GetSingleProvider(FileSystemProvider.ProviderName);
+                        return (fileSystemProvider.Home + path.Substring(1)).Replace(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar);
+                    }
+                    catch
+                    {
+                        return path;
+                    }
+                }
+                else
+#endif
+                if (context.SessionState.Path.IsPSAbsolute(path, out driveName))
+                {
+#if !UNIX
+                    // check if the driveName is an actual disk drive on Windows, if so, no expansion
+                    foreach (var drive in DriveInfo.GetDrives())
```

Comment:
Do we know how expensive this is? Is it possible to simplify it to simply testing if the drive name is just a single alpha char?

Reply:
A single char drive doesn't have to be a actual disk drive, this is checking if it's a logical drive otherwise expanding it results in the incorrect path.